name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/DekaBankPDFExtractor.java

```diff
@@ -924,6 +929,29 @@ else if (currentSecurityItem != null && currentSecurityItem == securityItem)
                                         t.setAmount(asAmount(v.get("amount")));
                                         t.setCurrencyCode(CurrencyUnit.EUR);
 
+                                        // @formatter:off
+                                        // Deka indicates only up to 3 digits after the decimal point.
+                                        // If the purchase or sale of shares is smaller, then we calculate the shares.
+                                        //
+                                        // Verkauf / Tausch Deka-Renten konservativ 0,02 48,370000 +0,000 09.06.2020 09.06.2020
+                                        // Kauf / Tausch Deka-RentenStrategie 0,02 91,710000 +0,000 09.06.2020 09.06.2020
+                                        //
+                                        // @formatter:on
+                                        if (t.getPortfolioTransaction().getShares() == 0 && t.getPortfolioTransaction().getAmount() != 0)
+                                        {
+                                            if ("Verkauf / Tausch".equals(trim(v.get("note"))))
+                                                t.setType(PortfolioTransaction.Type.SELL);
+
+                                            BigDecimal amountPerShare = BigDecimal.valueOf(asAmount(v.get("amountPerShare")));
+                                            BigDecimal amount = BigDecimal.valueOf(asAmount(v.get("amount")));
+
+                                            int sharesPrecision = Values.Share.precision() * 2;
+                                            BigDecimal shares = amount.divide(amountPerShare, sharesPrecision, RoundingMode.HALF_UP);
+
+                                            t.setNote(Messages.PDFLabelShares + ": " + v.get("shares") + " \u2259 "
+                                                            + shares.setScale(8, RoundingMode.HALF_UP));
```

Comment:
This code does not change any values on the imported transactions (besides adding a comment to the transaction). Shouldn't it call t.setShares()?

And then the code first rounds the division to 16 decimal places only to immediately round it to 8 places. You can skip the first rounding then.

And as long it is just a note, I do not have a problem with the hard-coded scale of 8. But if we use this value somewhere, it should use the precision from the values.

Reply:
I understand.
Done.
A note is no longer set, but the shares are calculated and set.

Reply:
I noticed there is no test case that is going into that branch of code to "fix" the number of shares (I wanted to change the BigInteger -> double -> long conversion because why introduce the low precision double in the middle).

Do you have an example? Or we build an example to safeguard this piece of code?

(I'll merge this change anyway because it makes sense to get the updates to DEKA)

Reply:
Yes, sure -> [here](https://github.com/buchen/portfolio/blob/e716242a51e0f7126a4f989476bc4b5cc1d3c0eb/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/dekabank/DekaBankPDFExtractorTest.java#L20892:L20905)